### PR TITLE
Add Pattgen to EarlGrey

### PR DESCRIPTION
**Pull request by martin-lueker**
Initial attempt at adding Pattgen.   Added to Peri xbar.

TODO: connect interrupts & pinmux.

### CODE REVIEW

**Comment by martin-lueker** [comment]
@tunghoang290780: There is a change to pattgen/dv/tb.sv in here.   Could you take a quick look?

**Comment by tjaychen** [comment]
this LGTM from `top_earlgrey.hjson` side.

**Comment by martin-lueker** [comment]
Thanks Tim!
Since there's a lot of changes here, I'm temporarily suppressing the autogen code to make the changes clearer.   This will pass CI once I add the autogen code back.

**Comment by tunghoang290780** [comment]
@martin-lueker LGTM with changes in tb.sv

**Review by tunghoang290780 — APPROVED** [review]

**Comment by msfschaffner** [comment]
> Thanks Tim!
> Since there's a lot of changes here, I'm temporarily suppressing the autogen code to make the changes clearer. This will pass CI once I add the autogen code back.

@martin-lueker, the way we usually handle this in a PR is to move all autogen'd changes to a second commit. that way, reviewers can just click on the first commit to review the changes without any clutter, and CI would still run on the full change-set.

**Comment by martin-lueker** [comment]
> > Thanks Tim!
> > Since there's a lot of changes here, I'm temporarily suppressing the autogen code to make the changes clearer. This will pass CI once I add the autogen code back.
> 
> @martin-lueker, the way we usually handle this in a PR is to move all autogen'd changes to a second commit. that way, reviewers can just click on the first commit to review the changes without any clutter, and CI would still run on the full change-set.

Hi @msfschaffner, That makes sense, and in fact that's what I had done originally. (!)  I've put them back now.   Feel free to take a look. -M

**Review by msfschaffner — APPROVED** [review]
Thanks @martin-lueker, this LGTM. Is this ready to be merged from your side?

File: hw/ip/pattgen/rtl/pattgen.sv
Comment (imphil) [inline hw/ip/pattgen/rtl/pattgen.sv:32]: nit: missing space before/after the `=`.

**Review by imphil — APPROVED** [review]
LGTM, thanks!

**Comment by msfschaffner** [comment]
@martin-lueker can you address @imphil's comments? once that is done, we can move ahead and merge this in.

**Comment by martin-lueker** [comment]
> @martin-lueker can you address @imphil's comments? once that is done, we can move ahead and merge this in.

Yes certainly.  (Sorry for the delay, I've been OOO the last couple weeks).

**Review by lenary — APPROVED** [review]
Software Changes LGTM.

File: hw/ip/pattgen/dv/tb.sv
Comment (martin-lueker) [reply hw/ip/pattgen/dv/tb.sv:44]: Check.

File: hw/ip/pattgen/rtl/pattgen.sv
Comment (martin-lueker) [reply hw/ip/pattgen/rtl/pattgen.sv:32]: Check.

Comment (martin-lueker) [reply hw/ip/pattgen/rtl/pattgen.sv:32]: Thanks @imphil.   Done.

File: hw/ip/pattgen/dv/tb.sv
Comment (martin-lueker) [reply hw/ip/pattgen/dv/tb.sv:44]: Thanks @imphil.  Done.

**Comment by martin-lueker** [comment]
Thanks, everyone.  @imphil, your changes are made, and it has been rebased.  Just waiting for CI now.

**Comment by mateenfaisal** [comment]
as Pattgen is added to Earl Grey, @martin-lueker can this issue be closed?

**Comment by martin-lueker** [comment]
Yes.

**Comment by mateenfaisal** [comment]
Ah, sorry I meant can we close this issue https://github.com/lowRISC/opentitan/issues/4156

**Pull request merged**
